serializers/src/main/java/io/pravega/schemaregistry/serializers/AbstractDeserializer.java

```diff
@@ -0,0 +1,116 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.serializers;
+
+import com.google.common.base.Preconditions;
+import io.pravega.schemaregistry.client.SchemaRegistryClient;
+import io.pravega.schemaregistry.contract.data.EncodingId;
+import io.pravega.schemaregistry.contract.data.EncodingInfo;
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.schemas.Schema;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.Nullable;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+@Slf4j
+abstract class AbstractDeserializer<T> extends BaseDeserializer<T> {
+    private static final int HEADER_SIZE = 1 + Integer.BYTES;
+
+    private final String groupId;
+    private final SchemaRegistryClient client;
+    // This can be null. If no schema is supplied, it means the intent is to deserialize into writer schema. 
+    // If headers are not encoded, then this will be the latest schema from the registry
+    private final SchemaInfo schemaInfo;
+    private final boolean encodeHeader;
+    private final SerializerConfig.Decoder decoder;
+    private final boolean skipHeaders;
+    private final EncodingCache encodingCache;
+
+    protected AbstractDeserializer(String groupId,
+                                   SchemaRegistryClient client,
+                                   @Nullable Schema<T> schema,
+                                   boolean skipHeaders,
+                                   SerializerConfig.Decoder decoder,
+                                   EncodingCache encodingCache,
+                                   boolean encodeHeader) {
+        Preconditions.checkNotNull(groupId);
+        Preconditions.checkNotNull(client);
+        Preconditions.checkNotNull(encodingCache);
+        this.groupId = groupId;
+        this.client = client;
+        this.encodingCache = encodingCache;
+        this.schemaInfo = schema == null ? null : schema.getSchemaInfo();
+        this.encodeHeader = encodeHeader;
+        this.skipHeaders = skipHeaders;
+        this.decoder = decoder;
+            
+        initialize();
+    }
+
+    private void initialize() {
+        if (schemaInfo != null) {
+            log.info("Validate caller supplied schema.");
+            if (!client.canReadUsing(groupId, schemaInfo)) {
+                throw new IllegalArgumentException("Cannot read using schema" + schemaInfo.getType());
+            }
+        } else {
+            if (!this.encodeHeader) {
+                log.warn("No reader schema is supplied and stream does not have encoding headers.");
+            }
+        }
+    }
+    
+    @SneakyThrows(IOException.class)
+    @Override
+    public T deserialize(ByteBuffer data) {
+        int start = data.arrayOffset() + data.position();
```

Comment:
hmm i had made this assumption because this buffer is given to the serializer by pravega reader client as it reads some data from pravega and asks the deserializer to deserialize it. 

if the buffer has no array backing it, i have changed the code to return null. is that acceptable?